### PR TITLE
Add USSD registration app

### PR DESCRIPTION
**Pull request by NicolaasDieBaas**

### CODE REVIEW

**Comment by NicolaasDieBaas** [comment]
@gsvr Please review / gooi 'n ogie asseblief ;)

File: src/ussd_registration.js [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+                "Please select the month the woman had her last period:",
+            "state_last_period_day":
+                "What day of the month did the woman start her last period? For example, 12.",
+            "state_baby_birth_date":
+                "Select the month & year the baby was born:",
+            "state_baby_birth_day":
+                "What day of the month was the baby born? For example, 12.",
+            "state_msg_language":
+                "Which language would this person like to receive these messages in?",
+            "state_msg_call_or_text":
```
Comment (gsvr) [inline src/ussd_registration.js]: Let's call this 'state_msg_type' for better corellation with the voice_registration app

File: src/ussd_registration.js [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+                "How would this person like to get messages?",
+            "state_receive_calls_days":
```
Comment (gsvr) [inline src/ussd_registration.js]: Let's call this 'state_voice_days'

File: src/ussd_registration.js [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+                "We will call them twice a week. On what days would the person like to receive these calls?",
+            "state_receive_calls_time":
```
Comment (gsvr) [inline src/ussd_registration.js]: Let's call this 'state_voice_times'

File: src/ussd_registration.js [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        };
+
+        get_sms_text = function(msg_receiver) {
+            return msg_receiver === 'time_out'
+                ? smss.time_out : null;
+        };
+
+        var errors = {
+            "state_auth_code":
+                "Sorry, that is not a valid number. Please enter your unique personnel code. For example, 12345."
```
Comment (gsvr) [inline src/ussd_registration.js]: Change 'number' to 'input'

File: src/ussd_registration.js [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+        };
+
+        get_error_text = function(name) {
+            return errors[name] || "Sorry, that is not a valid number. " + questions[name];
```
Comment (gsvr) [inline src/ussd_registration.js:78]: Change the string here to "Sorry, not a valid input. " so it's the same as FamilyConnect for now. We haven't received specific instructions for this project.

File: src/ussd_registration.js [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+            });
+        });
+
+        // ChoiceState st-03
+        self.add('state_msg_receiver', function(name) {
+            return new ChoiceState(name, {
+                question: $(questions[name]),
+                choices: [
+                    new Choice('the_mother', $("The Mother")),
+                    new Choice('the_father', $("The Father")),
```
Comment (gsvr) [inline src/ussd_registration.js]: Bike shedding - please just make it 'mother' and 'father', not 'the_mother' and 'the_father'

File: src/ussd_registration.js [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+            });
+        });
+
+        // ChoiceState st-04
+        self.add('state_pregnancy_status', function(name) {
+            return new ChoiceState(name, {
+                question: $(questions[name]),
+                choices: [
+                    new Choice('the_mother_pregnant', $("The mother is pregnant")),
+                    new Choice('the_mother_baby', $("The mother has a baby under 1 year old"))
```
Comment (gsvr) [inline src/ussd_registration.js]: You can use the `return choice.value;` here, or just change the values to 'pregnant' and 'baby'

File: src/ussd_registration.js [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        });
+
+        // ChoiceState st-09
+        self.add('state_receive_calls_days', function(name) {
+            return new ChoiceState(name, {
+                question: $(questions[name]),
+                error: $(get_error_text(name)),
+                choices: [
+                    new Choice('monday_and_wednesday', $('Monday and Wednesday')),
+                    new Choice('tuesday_and_thursday', $('Tuesday and Thursday'))
```
Comment (gsvr) [inline src/ussd_registration.js]: make these choice.values `mon_wed` and `tue_thu`

File: src/ussd_registration.js [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+        });
+
+        // ChoiceState st-08
+        self.add('state_msg_call_or_text', function(name) {
+            return new ChoiceState(name, {
+                question: $(questions[name]),
+                error: $(get_error_text(name)),
+                choices: [
+                    new Choice('voice_calls', $('Voice calls')),
+                    new Choice('text_smss', $('Text SMSs'))
```
Comment (gsvr) [inline src/ussd_registration.js]: Make these choice.values `voice` and `sms`

File: src/ussd_registration.js [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+        });
+
+        // ChoiceState st-10
+        self.add('state_receive_calls_time', function(name) {
+            return new ChoiceState(name, {
+                question: $(questions[name]),
+                error: $(get_error_text(name)),
+                choices: [
+                    new Choice('between_9_11am', $('Between 9-11am')),
+                    new Choice('between_2_5pm', $('Between 2-5pm'))
```
Comment (gsvr) [inline src/ussd_registration.js]: Make these `9_11` and `2_5`

File: src/ussd_registration.js [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                ],
+                next: 'state_end_thank_you_calls'
+            });
+        });
+
+        // EndState st-11
+        self.add('state_end_thank_you_calls', function(name) {
```
Comment (gsvr) [inline src/ussd_registration.js]: Just call this `state_end_voice`

File: src/ussd_registration.js [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+        });
+
+        // PaginatedChoiceState st-12
+        self.add('state_baby_birth_date', function(name) {
+            var today = go.utils.get_today(self.im.config);
+            var start_month = today.month();
+            return new PaginatedChoiceState(name, {
+                question: $(questions[name]),
+                error: $(get_error_text(name)),
+                characters_per_page: 160,
```
Comment (gsvr) [inline src/ussd_registration.js]: Let's set this to 182 for now to line up with the test harness. No info yet on what it needs to be.

File: src/ussd_registration.js [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+                    } else {
+                        return $(get_error_text(name));
+                    }
+                },
+                next: 'state_msg_language'
+            });
+        });
+
+        // EndState st-15
+        self.add('state_end_thank_you_texts', function(name) {
```
Comment (gsvr) [inline src/ussd_registration.js]: Just call this `state_end_sms`

File: test/ussd_registration.test.js [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+            });
+            it("to state_last_period_month", function() {
+                return tester
+                    .setup.user.addr('082111')
+                    .inputs(
+                        {session_event: 'new'}  // dial in
+                        , '12345'   // state_auth_code - personnel code
+                        , '0803304899' // state_msisdn - mobile number
+                        , '1'  // state_msg_receiver - mother
+                        , '1'  // state_msg_pregnant - mother
```
Comment (gsvr) [inline test/ussd_registration.test.js]: I think `mother` here should be `pregnant` and it should be `state_pregnancy_status`

File: test/ussd_registration.test.js [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+                    .check.interaction({
+                        state: 'state_end_thank_you_calls',
+                        reply: "Thank you. The person will now start receiving calls on [day and day] between [time - time]."
+                    })
+                    .run();
+            });
+        });
+
+        // flow testing case 2 - mother has baby and opts for text sms's
+        describe("Flow testing - mother with baby, text sms's", function() {
```
Comment (gsvr) [inline test/ussd_registration.test.js]: I think I may have half-answered one of your questions - let's discuss testing duplication face to face.

File: test/ussd_registration.test.js [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+            });
+            it("to state_last_period_month", function() {
+                return tester
+                    .setup.user.addr('082111')
+                    .inputs(
+                        {session_event: 'new'}  // dial in
+                        , '12345'   // state_auth_code - personnel code
+                        , '0803304899' // state_msisdn - mobile number
+                        , '1'  // state_msg_receiver - mother
+                        , '1'  // state_msg_pregnant - mother
```
Comment (gsvr) [inline test/ussd_registration.test.js:408]: This line should be `, '2'  // state_pregnancy_status - baby`, methinks

File: test/ussd_registration.test.js [hunk 17]
```diff
@@ -0,0 +1,578 @@
...
+                        , '3'  // state_last_period_month - May 15
+                        , '12' // state_last_period_day - 12
+                        , '1'  // state_msg_language - english
```
Comment (gsvr) [inline test/ussd_registration.test.js:459]: Choose a different language here

File: test/ussd_registration.test.js [hunk 18]
```diff
@@ -0,0 +1,578 @@
...
+                        {session_event: 'new'}  // dial in
+                        , 'aaaaa'  // state_auth_code - invalid personnel code
+                    )
+                    .check.interaction({
+                        state: 'state_auth_code',
+                        reply: "Sorry, that is not a valid number. Please enter your unique personnel code. For example, 12345."
+                    })
+                    .run();
+            });
+          /*  it("validate state_msg_receiver", function() {
```
Comment (gsvr) [inline test/ussd_registration.test.js]: Let's make this work rather than comment it out.

**Comment by gsvr** [comment]
@NicolaasDieBaas First pass review done.

**Comment by gsvr** [comment]
:+1: when travis is happy

**Pull request merged**